libraries/testing/tester.cpp

```diff
@@ -508,21 +516,23 @@ namespace eosio::testing {
       } );
 
       control->commit_block(br);
-      last_produced_block[producer_name] = control->head_block_id();
+      signed_block_ptr sb = control->head_block();
+      last_produced_block[producer_name] = sb->calculate_id();
 
-      _wait_for_vote_if_needed(*control);
+      _check_for_vote_if_needed(*control, sb);
 
-      return control->head_block();
+      return sb;
    }
 
-   void base_tester::_wait_for_vote_if_needed(controller& c) {
-      if (c.can_vote_on(c.head_block())) {
-         // wait for this node's vote to be processed
-         size_t retrys = 500;
-         while (!c.node_has_voted_if_finalizer(c.head_block_id()) && --retrys) {
-            std::this_thread::sleep_for(std::chrono::milliseconds(10));
-         }
-         FC_ASSERT(retrys, "Never saw this nodes vote processed before timeout");
+   void base_tester::_check_for_vote_if_needed(controller& c, const signed_block_ptr& b) {
+      if (_expect_votes && c.can_vote_on(b)) {
+         // `_expect_votes` should be true *only* when we expect an active finalizer to
+         // vote on every block.
+         // This is not the case for tests with forks, so for these tests we should set
+         // `_expect_votes` to false by calling `base_tester::do_check_for_votes(false)`
+         // ----------------------------------------------------------------------------
+         auto block_id = b->calculate_id();
+         FC_ASSERT(c.node_has_voted_if_finalizer(block_id), "Missing expected vote");
```

Comment:
done.

> The only tricky part would be getting getting the `block_handle` in `_finish_block`.

I added a `block_handle controller::head() const { return my->chain_head; }` member in `controller`.

Now that we have a `head()` member returning a `block_handle`, I would be in favor of removing the other accessors below it in the following link: https://github.com/AntelopeIO/spring/blob/df1c5ac8d7059ef05c1bd7feaff9bfefcc4862eb/libraries/chain/controller.cpp#L5095-L5118

Reply:
Looks great!

I think I agree with removing the other accessors which are made redundant by simply providing `head()`.
My only hesitation is that `head()` exposes the `block_handle` to components other than the internals of libchain and tester, e.g. plugins, and that wouldn't necessarily be a problem as long as it does not expose the internal `block_header_state`.

But `block_handle::internal()` return the internal `_bsp`. Maybe the name is sufficient to indicate to others that this is not part of the API that they should rely on. But maybe there is a way to use `protected` and `friend` to enforce better restrictions so that the compiler only allows the the code in `controller.cpp` to access the internal block header state via the `block_handle`?

Reply:
> But maybe there is a way to use protected and friend to enforce better restrictions so that the compiler only allows the the code in controller.cpp to access the internal block header state via the block_handle?

Just implemented this in latest commit https://github.com/AntelopeIO/spring/pull/316/commits/978e88864dcab02eecf7de504f2f2e9fdace3465